programs/stake/src/stake_state.rs

```diff
@@ -842,6 +823,94 @@ impl<'a> StakeAccount for KeyedAccount<'a> {
+    let stake_amount = account.lamports()?.saturating_sub(meta.rent_exempt_reserve); // can't stake the rent
+    Ok(ValidatedDelegatedInfo { stake_amount })
+}
+
+/// After calling `validate_split_amount()`, this struct contains calculated values that are used
+/// by the caller.
+#[derive(Copy, Clone, Debug, Default)]
+struct ValidatedSplitInfo {
+    source_remaining_balance: u64,
+    destination_rent_exempt_reserve: u64,
+}
+
+/// Ensure the split amount is valid.  This checks the source and destination accounts meet the
+/// minimum balance requirements, which is the rent exempt reserve plus the minimum stake
+/// delegation, and that the source account has enough lamports for the request split amount.  If
+/// not, return an error.
+fn validate_split_amount(
+    source_account: &KeyedAccount,
+    destination_account: &KeyedAccount,
+    lamports: u64,
+    source_meta: &Meta,
+) -> Result<ValidatedSplitInfo, InstructionError> {
+    let source_lamports = source_account.lamports()?;
+
+    // Split amount has to be something
+    if lamports == 0 {
+        return Err(InstructionError::InsufficientFunds);
+    }
+
+    // Obviously cannot split more than what the source account has
+    if lamports > source_lamports {
+        return Err(InstructionError::InsufficientFunds);
+    }
+
+    // Verify that the source account still has enough lamports left after splitting:
+    // EITHER at least the minimum balance, OR zero (in this case the source
+    // account is transferring all lamports to new destination account, and the source
+    // account will be closed)
+    let source_minimum_balance = source_meta
+        .rent_exempt_reserve
+        .saturating_add(MINIMUM_STAKE_DELEGATION);
+    let source_remaining_balance = source_lamports.saturating_sub(lamports);
+    if source_remaining_balance == 0 {
+        // full amount is a withdrawal
+        // nothing to do here
+    } else if source_remaining_balance < source_minimum_balance {
+        // the remaining balance is too low to do the split
+        return Err(InstructionError::InsufficientFunds);
+    } else {
+        // all clear!
+        // nothing to do here
+    }
+
+    // Verify the destination account also meets the minimum balance requirements
+    // This must account for:
+    // 1. The destination account having a different rent exempt reserve due to data size changes
+    // 2. The destination account being prefunded, which would lower the minimum split amount
+    let destination_rent_exempt_reserve = calculate_split_rent_exempt_reserve(
+        source_meta.rent_exempt_reserve,
+        source_account.data_len()? as u64,
+        destination_account.data_len()? as u64,
+    );
+    let destination_minimum_balance =
+        destination_rent_exempt_reserve.saturating_add(MINIMUM_STAKE_DELEGATION);
+    let minimum_split_amount =
+        destination_minimum_balance.saturating_sub(destination_account.lamports()?);
```

Comment:
You can do this in a separate PR, but this could create a strange situation.  Imagine that I have an active stake with a lot of SOL, and I want to split some of it into another account.  If the destination account already has `MINIMUM_STAKE_DELEGATION + destination_rent_exempt_reserve - 1`, I could split 1 active lamport into it, which means that it would have a smaller delegation than `MINIMUM_STAKE_DELEGATION`.

So you might need to have `minimum_split_amount` resolve to `max(MINIMUM_STAKE_DELEGATION, destination_minimum_balance.saturating_sub(destination_account.lamports()?))` if the source account is active

Reply:
Yeah, that's true. This gets elided at present, since `MINIMUM_STAKE_DELEGATION == 1`. Great spot, Jon

Reply:
Ah yes! To my happiness I'm glad that I haven't inadvertently changed any existing behavior! To resolve this, I've added/updated tests over here: https://github.com/solana-labs/solana/pull/23456.

This way, if the minimum stake delegation is ever raised, the tests will fail until this is resolved. Tracking that in #23458.

Reply:
Done! Tests were merged, this PR has rebased. And since #23458 has been deleted, I've implemented your fix in this PR as well.